File: src/lib/tauri/backups.ts
import { invoke } from "@tauri-apps/api/core";

import type {
  BackupExportResult,
  BackupImportResult,
  BackupPreview,
  ExportConnectionBackupInput,
  ImportConnectionBackupInput,
  InspectConnectionBackupInput,
} from "@/types/backups";

const EXPORT_CONNECTION_BACKUP_COMMAND = "export_connection_backup";
const INSPECT_CONNECTION_BACKUP_COMMAND = "inspect_connection_backup";
const IMPORT_CONNECTION_BACKUP_COMMAND = "import_connection_backup";

export function exportConnectionBackup(input: ExportConnectionBackupInput) {
  return invoke<BackupExportResult>(EXPORT_CONNECTION_BACKUP_COMMAND, { input });
}

export function inspectConnectionBackup(input: InspectConnectionBackupInput) {
  return invoke<BackupPreview>(INSPECT_CONNECTION_BACKUP_COMMAND, { input });
}

export function importConnectionBackup(input: ImportConnectionBackupInput) {
  return invoke<BackupImportResult>(IMPORT_CONNECTION_BACKUP_COMMAND, { input });
}

File: src-tauri/src/commands/backups.rs
use tauri::State;

use crate::models::backups::{
    BackupExportResultDto, BackupImportResultDto, BackupPreviewDto, ExportConnectionBackupInput,
    ImportConnectionBackupInput, InspectConnectionBackupInput,
};
use crate::models::error::CommandError;
use crate::services::backups::ConnectionBackupService;

#[tauri::command]
pub async fn export_connection_backup(
    input: ExportConnectionBackupInput,
    service: State<'_, ConnectionBackupService>,
) -> Result<BackupExportResultDto, CommandError> {
    let (path, password, include_credentials) = input.into_parts();
    service
        .export(path, password, include_credentials)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn inspect_connection_backup(
    input: InspectConnectionBackupInput,
    service: State<'_, ConnectionBackupService>,
) -> Result<BackupPreviewDto, CommandError> {
    let (path, password) = input.into_parts();
    service
        .inspect(path, password)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn import_connection_backup(
    input: ImportConnectionBackupInput,
    service: State<'_, ConnectionBackupService>,
) -> Result<BackupImportResultDto, CommandError> {
    let (path, password, conflict_strategy) = input.into_parts();
    service
        .import(path, password, conflict_strategy)
        .await
        .map(Into::into)
        .map_err(Into::into)
}
